src/stream/xml_reader.rs

```diff
@@ -1,221 +1,262 @@
+                            }
+                        }
+
                         _ => return Err(self.with_pos(ErrorKind::UnknownXmlElement)),
                     }
                 }
-                Ok(XmlEvent::EndElement { name, .. }) => {
+                XmlEvent::End(name) => {
                     // Check the corrent element is being closed
                     match self.element_stack.pop() {
-                        Some(ref open_name) if &name.local_name == open_name => (),
-                        Some(ref _open_name) => {
+                        Some(open_name) if name.local_name() != open_name.as_ref() => {
                             return Err(self.with_pos(ErrorKind::UnclosedXmlElement))
                         }
                         None => return Err(self.with_pos(ErrorKind::UnpairedXmlClosingTag)),
+                        _ => {}
                     }
 
-                    match &name.local_name[..] {
-                        "array" | "dict" => return Ok(Some(Event::EndCollection)),
-                        "plist" | _ => (),
-                    }
-                }
-                Ok(XmlEvent::EndDocument) => {
-                    if self.element_stack.is_empty() {
-                        return Ok(None);
-                    } else {
-                        return Err(self.with_pos(ErrorKind::UnclosedXmlElement));
+                    match name.local_name() {
+                        b"array" | b"dict" => return Ok(Some(Event::EndCollection)),
+                        b"plist" | _ => (),
                     }
                 }
-
-                Ok(XmlEvent::Characters(c)) => {
-                    if !is_whitespace_str(&c) {
-                        return Err(
-                            self.with_pos(ErrorKind::UnexpectedXmlCharactersExpectedElement)
-                        );
-                    }
+                XmlEvent::Eof if self.element_stack.is_empty() => return Ok(None),
+                XmlEvent::Eof => return Err(self.with_pos(ErrorKind::UnclosedXmlElement)),
+                XmlEvent::Text(_) => {
+                    return Err(self.with_pos(ErrorKind::UnexpectedXmlCharactersExpectedElement))
                 }
-                Ok(XmlEvent::CData(_)) | Ok(XmlEvent::Comment(_)) | Ok(XmlEvent::Whitespace(_)) => {
-                    unreachable!("parser does not output CData, Comment or Whitespace events")
+                XmlEvent::PI(_)
+                | XmlEvent::Decl(_)
+                | XmlEvent::DocType(_)
+                | XmlEvent::CData(_)
+                | XmlEvent::Comment(_) => {
+                    // skip
                 }
-                Ok(XmlEvent::ProcessingInstruction { .. }) => (),
-                Err(err) => return Err(from_xml_error(err)),
             }
         }
     }
-
-    fn with_pos(&self, kind: ErrorKind) -> Error {
-        kind.with_position(convert_xml_pos(self.xml_reader.position()))
-    }
 }
 
 impl<R: Read> Iterator for XmlReader<R> {
```

Comment:
I know I'm being a bit picky, but can this go above the impl block for `ReaderState`.

Reply:
Moved it up.